storage/innobase/buf/buf0dblwr.cc

```diff
@@ -365,25 +366,17 @@ void buf_dblwr_t::recover()
   for (recv_dblwr_t::list::iterator i= recv_sys.dblwr.pages.begin();
        i != recv_sys.dblwr.pages.end(); ++i, ++page_no_dblwr)
   {
-    byte *page= *i;
-    const uint32_t page_no= page_get_page_no(page);
+    const uint32_t page_no= page_get_page_no(*i);
     if (!page_no) /* recovered via recv_dblwr_t::restore_first_page() */
       continue;
 
-    const lsn_t lsn= mach_read_from_8(page + FIL_PAGE_LSN);
-    if (recv_sys.parse_start_lsn > lsn)
-      /* Pages written before the checkpoint are not useful for recovery. */
+    const lsn_t lsn= mach_read_from_8(*i + FIL_PAGE_LSN);
+    if (recv_sys.parse_start_lsn > lsn || lsn > recv_sys.scanned_lsn)
+      /* Pages written before or after the recovery range are not usable. */
       continue;
```

Comment:
I realized that `recv_sys.scanned_lsn` at this point is not guaranteed to be the final one. We will only have read enough log so that `recv_validate_tablespace()` will succeed. We must ensure that all log has been parsed up to this point.

Starting with a5a2ef079cec378340d8b575aef05974b0b3442e there actually shouldn’t be any duplicate page numbers in the doublewrite buffer, except when the tail of an earlier `buf_dblwr_t::flush_buffered_writes()` had not been overwritten by subsequent smaller `buf_dblwr_t::flush_buffered_writes()`.

This logic change is fixing a bug that we’d attempt to recover a too new copy of a page from the doublewrite buffer. But, because the log parsing is not necessarily completed at this point, it could happen that we fail to recover a valid data directory because we would ignore a valid (not too new) doublewrite copy of a page.

I’m starting to think that we should ensure that `log_sys.lsn` is correct at this point of execution. This will require some more work.

Reply:
I added an assertion `recv_sys.scanned_lsn == log_sys.get_lsn()` at the start of `buf_dblwr_t::recover()`. It does not fail. It turns out that in `recv_recovery_from_checkpoint_start()` the call `buf_dblwr.recoevr()` is preceded by the call `log_sys.set_lsn(recv_sys.recovered_lsn)`. So, we indeed do know the final recovered LSN upon executing the recovery from the doublewrite buffer.